cpp/src/arrow/ipc/json_internal.cc

```diff
@@ -105,10 +106,25 @@ class SchemaWriter {
       RETURN_NOT_OK(VisitField(field));
     }
     writer_->EndArray();
+    WriteKeyValueMetadata(schema_.metadata());
     writer_->EndObject();
     return Status::OK();
   }
 
+  void WriteKeyValueMetadata(const std::shared_ptr<const KeyValueMetadata>& metadata) {
+    if (metadata == nullptr) {
+      return;
+    }
+    writer_->Key("metadata");
+
+    writer_->StartObject();
+    for (int64_t i = 0; i < metadata->size(); ++i) {
+      writer_->Key(metadata->key(i).c_str());
+      writer_->String(metadata->value(i).c_str());
+    }
```

Comment:
This make the assumption that the keys are unique (not disallowed by the format even if advisable). Would a list of `{'key': ..., 'value': ...}` be better?

Reply:
Aside: in the JavaScript data world (e.g. for D3 users) there is a mantra I've heard "don't store data as object attribute names"

Reply:
@wesm I can change this, but it's worth mentioning that what I've written is how Java (and incompletely: Go, JS) serialize metadata. Furthermore, key value entries are compared in key sorted order in C++ (so key value reordering does not matter). Is there a use case for duplicated metadata keys? It seems we might prefer to require that they be unique, since all reference implementations currently do. Either way, this probably warrants a mailing list discussion.

Reply:
Per mailing list discussion, I'll change to use a list of keyvalues instead of a dict and skip Java as well as Go/JS